diff --git a/app/build.gradle b/app/build.gradle
index 167ae45..652eb33 100644
--- a/app/build.gradle
+++ b/app/build.gradle
@@ -11,6 +11,25 @@ android {
         versionCode 42
         versionName "1.7.1"
         archivesBaseName = "XPrivacyLua-v$versionName-$versionCode"
+
+        externalNativeBuild {
+            // https://developer.android.com/ndk/guides/cmake.html
+            cmake {
+                cppFlags "-O2 -Wall -DLUA_COMPAT_ALL -D\"getlocaledecpoint() ='.'\""
+                arguments "-DANDROID_PLATFORM=android-23"
+            }
+        }
+
+        ndk {
+            // https://developer.android.com/ndk/guides/abis.html#sa
+            abiFilters 'armeabi-v7a', 'arm64-v8a', 'x86', 'x86_64'
+        }
+    }
+
+    externalNativeBuild {
+        cmake {
+            path "CMakeLists.txt"
+        }
     }
 
     buildTypes {
diff --git a/app/src/main/java/eu/faircode/xlua/XLua.java b/app/src/main/java/eu/faircode/xlua/XLua.java
index 41c9e81..8aeaaa2 100644
--- a/app/src/main/java/eu/faircode/xlua/XLua.java
+++ b/app/src/main/java/eu/faircode/xlua/XLua.java
@@ -47,6 +47,7 @@ import org.luaj.vm2.lib.jse.CoerceJavaToLua;
 import org.luaj.vm2.lib.jse.JsePlatform;
 
 import java.io.ByteArrayInputStream;
+import java.io.File;
 import java.io.InputStream;
 import java.lang.reflect.Field;
 import java.lang.reflect.Method;
@@ -69,9 +70,13 @@ public class XLua implements IXposedHookZygoteInit, IXposedHookLoadPackage {
     private static final String TAG = "XLua.Xposed";
 
     private static int version = -1;
+    private static String apk;
+
+    static native void test();
 
     public void initZygote(final IXposedHookZygoteInit.StartupParam startupParam) throws Throwable {
         Log.i(TAG, "initZygote system=" + startupParam.startsSystemServer + " debug=" + BuildConfig.DEBUG);
+        apk = startupParam.modulePath;
     }
 
     public void handleLoadPackage(final XC_LoadPackage.LoadPackageParam lpparam) throws Throwable {
@@ -79,8 +84,19 @@ public class XLua implements IXposedHookZygoteInit, IXposedHookLoadPackage {
         String self = XLua.class.getPackage().getName();
         Log.i(TAG, "Loaded " + lpparam.packageName + ":" + uid);
 
-        if ("android".equals(lpparam.packageName))
+        if ("android".equals(lpparam.packageName)) {
+            try {
+                String libname = new File(apk).getParent() + "/lib/arm64/liblua.so";
+                Log.i(TAG, "Loading Lua file=" + libname);
+                System.load(libname);
+                Log.i(TAG, "Loaded Lua");
+                test();
+                Log.i(TAG, "Tested Lua");
+            } catch (Throwable ex) {
+                Log.w(TAG, ex.getMessage());
+            }
             hookAndroid(lpparam);
+        }
 
         if ("com.android.providers.settings".equals(lpparam.packageName))
             hookSettings(lpparam);
@@ -419,6 +435,13 @@ public class XLua implements IXposedHookZygoteInit, IXposedHookLoadPackage {
                                 // Execute hook
                                 private void execute(MethodHookParam param, String function) {
                                     try {
+                                        Log.i(TAG, "Testing Lua");
+                                        test();
+                                        Log.i(TAG, "Tested Lua");
+                                    } catch (Throwable ex) {
+                                        Log.e(TAG, ex.getMessage());
+                                    }
+                                    try {
                                         long run = SystemClock.elapsedRealtime();
 
                                         // Initialize Lua runtime
